packages/runtime/src/useCases/common/validation/SchemaValidator.ts

```diff
@@ -1,17 +1,34 @@
 import { ErrorObject } from "ajv";
+import { DateTime } from "luxon";
 import { RuntimeErrors } from "../RuntimeErrors";
 import { JsonSchema, JsonSchemaValidationResult } from "../SchemaRepository";
 import { IValidator } from "./IValidator";
 import { ValidationFailure } from "./ValidationFailure";
 import { ValidationResult } from "./ValidationResult";
 
-export class SchemaValidator<T> implements IValidator<T> {
+export class SchemaValidator<T extends object> implements IValidator<T> {
     public constructor(protected readonly schema: JsonSchema) {}
 
     public validate(input: T): ValidationResult {
-        const validationResult = this.schema.validate(input);
+        const validationResult = this.convertValidationResult(this.schema.validate(input));
 
-        return this.convertValidationResult(validationResult);
+        if ("expiresAt" in input && input.expiresAt) {
+            if (DateTime.fromISO(input.expiresAt as string) <= DateTime.utc()) {
+                validationResult.addFailure(new ValidationFailure(RuntimeErrors.general.invalidPropertyValue(`'expiresAt' must be in the future`), "expiresAt"));
+            }
+        }
+
+        if ("passwordProtection" in input && input.passwordProtection) {
+            const passwordProtection = input.passwordProtection as { password: string; passwordIsPin?: true };
+
+            if (passwordProtection.passwordIsPin) {
+                if (!/^[0-9]{4,16}$/.test(passwordProtection.password)) {
+                    validationResult.addFailure(new ValidationFailure(RuntimeErrors.general.invalidPin()));
+                }
+            }
+        }
+
+        return validationResult;
```

Comment:
What's going on here? Why do you put logic for some usescases in the baseclass?! It would be way cleaner if you just extend the SchemaValidator and put it there any only use the new extended version where required.